DOCS-#73: Add documentation page on CLI usage (#83)
modin-project/unidist: docs/using_cli.rst

..
      Copyright (C) 2021-2022 Modin authors

      SPDX-License-Identifier: Apache-2.0

Unidist Command Line Interface
""""""""""""""""""""""""""""""

unidist provides its own command line interface (CLI) for running python applications.

> The CLI is a recommended way for executing program with unidist. 

Author: fixed

The CLI can be used in two ways. If unidist is installed in user's environment (from conda, pip or sources) the user
has *unidist* executable binary. In this case a recommended way to run an application is the following:

.. code-block:: bash

   $ unidist script.py

In the case of using unidist sources, the application can be run as follows:

.. code-block:: bash

>    $ python unidist/cli script.py  # The running is happened from the unidist's root directory

Author: fixed

CLI Options
===========

> unidist behavior can be tuned via CLI options. All actual supported options can always be obtained using ``--help``:

Author: fixed

                           specify a number of CPUs per node used by the backend in a cluster. Can accept multiple
                           values in the case of running in the cluster. Default value is taken from
                           'UNIDIST_CPUS' environment variable. If 'UNIDIST_CPUS' isn't set, value is equal to the
                           number of CPUs on a head node.
   -hosts HOSTS [HOSTS ...], --hosts HOSTS [HOSTS ...]
                           specify node(s) IP address(es) to use by the backend. Can accept multiple values in the
                           case of running in a cluster. Default is 'localhost'.

   required arguments:
   script                specify a script to be run

   Ray backend-specific arguments:
   -redis_pswd REDIS_PASSWORD, --redis_password REDIS_PASSWORD
                           specify redis password to connect to existing Ray cluster.

Usage Examples
==============

* Default is running the script with unidist on Ray backend in a single node:

  .. code-block:: bash

      $ unidist script.py

* Running the script with unidist on Dask backend via ``pytest`` with a pytest-specific option:

  .. code-block:: bash

>       $ unidist -m pytest script.py -b dask --verbose  # --verbose is pytest-specific option

Author: fixed

..
      Copyright (C) 2021-2022 Modin authors

      SPDX-License-Identifier: Apache-2.0

Unidist Command Line Interface
""""""""""""""""""""""""""""""

Unidist provides its own command line interface (CLI) for running applications.

> The CLI is recommended for an executing unidist programs. 

Author: Applied excluding "your". I would prefer docs without reader mention.

   -hosts HOSTS [HOSTS ...], --hosts HOSTS [HOSTS ...]
                           specify node(s) IP address(es) to use by the backend. Can accept multiple values in the
                           case of running in a cluster. Default is 'localhost'.

   required arguments:
   script                specify a script to be run

   Ray backend-specific arguments:
   -redis_pswd REDIS_PASSWORD, --redis_password REDIS_PASSWORD
                           specify redis password to connect to existing Ray cluster.

Usage Examples
==============

Suppose that user has file ``script.py`` with unidist functionality. Consider several cases of running this script with
different options/environment variables combinations.

* Default is running in a single node mode on Ray backend:

  .. code-block:: bash

      $ unidist script.py

* Running ``pytest`` on Dask backend with a pytest-specific option:

  .. code-block:: bash

      $ unidist -m pytest script.py -b dask --verbose  # --verbose is pytest-specific option

  or (use ``UNIDIST_BACKEND`` environment variable instead of ``-b`` option):

  .. code-block:: bash

      $ export UNIDIST_BACKEND=Dask

>       $ unidist -m pytest script.py --verbose  # --verbose is pytest-specific option

Reviewer: ```suggestion
      $ unidist -m pytest script.py --verbose  # --verbose is the pytest-specific option
```

> * Running in a single node mode on MPI backend using 8 workers:

Reviewer: ```suggestion
* Running the script with unidist on MPI backend using 8 workers in a single node:
```

  .. code-block:: bash

      $ unidist script.py --backend mpi --num_cpus 8

> * Running the script on two nodes using MPI backend. Nodes will have 16 and 32 workers, respectively:

Reviewer: ```suggestion
* Running the script with unidist on MPI backend on two nodes. The nodes have 16 and 32 workers, respectively:
```

  .. code-block:: bash

      $ export UNIDIST_BACKEND=MPI
      $ unidist script.py -hosts localhost x.x.x.1 --num_cpus 16 32

> * Running the script on a Ray pre-initialized cluster:

Reviewer: ```suggestion
* Running the script with unidist on Ray backend with pre-initialized Ray cluster:
```

  .. code-block:: bash

>       $ unidist script.py -hosts x.x.x.1 -redis_pswd 123456 # x.x.x.1 is IP-address of a head node of Ray cluster

Reviewer: ```suggestion
      $ unidist script.py -hosts x.x.x.1 -redis_pswd 123456 # x.x.x.1 is the IP-address of a head node of the Ray cluster
```

> * Running the script on a Dask pre-initialized cluster:

Reviewer: ```suggestion
* Running the script with unidist on Dask backend with a pre-initialized Dask cluster:
```

  .. code-block:: bash

>       $ unidist script.py -b dask -hosts x.x.x.1:port # x.x.x.1:port is IP-address with port of a Dask-scheduler

Reviewer: ```suggestion
      $ unidist script.py -b dask -hosts x.x.x.1:port # x.x.x.1:port is the IP-address with the port of a dask-scheduler
```

.. note:: 

>     Currently, to use unidist with Ray and Dask backends on cluster, need to pre-initialize Dask/Ray cluster

Reviewer: ```suggestion
    Currently, in order to use unidist with Ray or Dask backend on a cluster, Ray/Dask cluster needs to be pre-initialized
```

>     using their own documentation (`Ray Guide <https://docs.ray.io/en/latest/starting-ray.html#starting-ray-via-the-cli-ray-start>`_

Reviewer: ```suggestion
    using its own documentation (`Ray Guide <https://docs.ray.io/en/latest/cluster/cloud.html#manual-ray-cluster-setup>`_
```

>     and `Dask Guide <https://docs.dask.org/en/latest/how-to/deploy-dask-clusters.html>`_).

Reviewer: ```suggestion
    and `Dask Guide <https://docs.dask.org/en/latest/how-to/deploy-dask/cli.html>`_).
```
